src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -784,17 +786,14 @@ public RecoveryState recoveryState() {
         return this.recoveryState;
     }
 
-    public void performRecoveryFinalization(boolean withFlush) throws ElasticsearchException {
+    public void finalizeRecovery(boolean withFlush) throws ElasticsearchException {
```

Comment:
I think that's safe to catch it (we only flush from the gateway and there can't be any other recoveries going on). I'll move it there. In general I wonder if we should just flush all the time, it can not hurt to trim the translog - but that's a different change.